_posts/2016-02-22-pm-emulation.md

```diff
@@ -0,0 +1,89 @@
+---
+title: How to emulate Persistent Memory
+author: Maciej Maciejewski
+layout: post
+---
+Data allocated with NVML is put to the virtual memory address space, and concrete ranges are relying on result of `mmap(2)` operation performed on the user defined files.
+Such files can exist on any storage media, however data consistency assurance embedded within NVML requires frequent synchronisation of data that is being modified.
+Depending on platform capabilities, and underlying device where the files are, a different set of commands is used to facilitate synchronisation.
+It might be `msync(2)` for the regular hard drives, or combination of cache flushing instructions followed by memory fence instruction for the real persistent memory.
+
+Although application adaptation to NVML usage, and ability to operate on persistent memory might be done by relying on regular hard drive, it is not recommended due to the performance hit coming from *msync(2)* operation.
+That is the reason to work either with the real equipment or emulated environment. Since persistent memory is not yet commonly available we do recommend setting up emulation system, that will speed up development, and testing of the application you are converting. In the following steps we shall cover how to setup such system.
+
+### Hardware and system requirements
+Emulation environment is available at the current stage only for Linux systems, and should work on any hardware or virtualized environment. Emulation of persistent memory is based on DRAM memory, that is seen by OS as Persistent Memory region. Due to being a DRAM based emulation it is very fast, but will likely loose all the data upon powering down the machine.
+It should as well work with any distribution able to handle official [kernel](https://www.kernel.org/).
+
+### Linux Kernel
+Download kernel sources from [official kernel pages](https://www.kernel.org/).
+Support for persistent memory devices and emulation is present in Kernel since 4.0 version, however it is recommended to use Kernel newer then 4.2 due to easier configuration of it. Following instruction relies on 4.2 or newer. Using Kernel older then 4.2 will require a bit more work to setup, and will not be described here. 
+Please note, that features and bug fixes around DAX support are being implemented as we speak, therefore it is recommended to use the newest stable Kernel if possible.
+To configure proper driver installation run `nconfig` and enable driver.
+
+{% highlight sh %}
+$ make nconfig
+	-> Device Drivers -> NVDIMM Support ->
+			<M>PMEM; <M>BLK; <*>BTT
+{% endhighlight %}
+
+You are ready to build your Kernel
+
+{% highlight sh %}
+$ make -jX
+	where X is the number of cores on the machine
+{% endhighlight %}
+
+Install the kernel
+
+{% highlight sh %}
```

Comment:
true, changed to 'console', it basically does no highlighting and looks normal, you can take a look at the result here:
http://maciuch.github.io/2016/02/22/pm-emulation.html